rfd/0XYZ-nested-accesslists.md

```diff
@@ -0,0 +1,121 @@
+---
+title: RFD 0XYZ - Nested Access lists
+authors: Alex McGrath (alex.mcgrath@goteleport.com)
+state: in development
+---
+
+# Related
+  * Partial implementation PR: https://github.com/gravitational/teleport/pull/38738
+
+# What
+
+This RFD proposes adding support for including owners and
+members in access lists from other access lists
+
+# Why
+
+This allows a hierarchical access list structure for organizations
+which is supported by access list implementations being integrated
+with, for example Azure Entra supports including groups in other
+groups which we want to be able to mimic the structure of in Teleport
+access lists.
+
+Users in an access list hierarchy will inherit the granted roles and
+traits for members and owners from access lists referencing the lists
+they're in.
+
+# Implementations
+
+New fields will be introduced into the access_list type:
+
+```yaml
+kind: access_list
+metadata:
+  name: e69aa529-2a7f-4de2-87be-97fd94309a9f
+spec:
+  grants:
+    roles:
+    - access
+    - auditor
+    traits: {}
+  # list of references to other access lists, for users to include in this access list
+  member_access_lists:
+  - name: ea4cbbc7-bee1-49b3-bf78-734b4b27ea38
+  # list of references to other access lists, for owners to include in this access list
+  owner_access_lists:
+  - name: 3e9df1e7-0b8a-4984-b2e8-5bc0d7b356a9
+  title: access-list-a
+version: v1
+```
+
+These fields will contain a list of other access lists that will be
+included in the access list.
+
+# Implementation considerations
+
+The implementation will not support looping within the heirarchy as
+this would introduce confusing options for configuration and will
+return an error if loops are introduced, as they dont have obvious
+real world use cases.
+
+# Examples
+
+
+User Alice is in the below access list which does not include any
+other access lists
+
+```yaml
+kind: access_list
+metadata:
+  name: acl-a
+spec:
+  grants:
+    roles:
+    - some-role
+    traits: {}
+
+  title: access-list-a
+version: v1
+```
+- `acl-c` includes members in `acl-a`, grants them `manager`
+- `acl-b` includes members in `acl-c`, and grants them `auditor` and `reviewer`
+
+```yaml
+kind: access_list
+metadata:
+  name: acl-b
+spec:
+  grants:
+    roles:
+    - auditor
+    - reviewer
+    traits: {}
+  # list of references to other access lists, for users to include in this access list
+  member_access_lists:
+  - name: acl-c
+  title: access-list-a
+version: v1
+---
+kind: access_list
+metadata:
+  name: acl-c
+spec:
+  grants:
+    roles:
+    - manager
+    traits: {}
+  # list of references to other access lists, for users to include in this access list
+  member_access_lists:
+  - name: acl-a
+  title: access-list-a
+version: v1
+
+```
+
+When calculating a users access roles, the tree of access lists will
```

Comment:
How does this affect the access request flow, will similar traversal happen there when calculating suggested access lists? For example, say there is an access list "Dev Team" that gives access to server S1. Then, say "Dev Team" contains another list "Core Team" that gives access to server S2.

Now Alice requests access to server S1. Will the request reviewer (if they're an owner of both lists) be able to promote her to either one or just the "Dev Team"? Can you add a section to the RFD that clarifies this.

Reply:
I'm not super experienced in how the access requests and access lists interact

In this case because alice is requesting server S1 which Dev team owners can give access to, is she actually requesting to be on the list or requesting the resource s1?

The way its implemented now, user permissions are calculated in IsAccessList(Member/Owner) and I have implemented it to just look up the tree in there, so its fairly transparent to access requests, as to whether a user is a member/owner via nesting or direct membership